src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetService.cs

```diff
@@ -0,0 +1,233 @@
+            .Where(rf => rf.Id == releaseFileId)
+            .Select(rf => new DataSet
+            {
+                Status = DataSetStatus.Draft,
+                Title = rf.Name!,
+                Summary = rf.Summary ?? "",
+                PublicationId = rf.ReleaseVersion.PublicationId,
+            })
+            .FirstAsync(cancellationToken);
+
+        publicDataDbContext.Add(dataSet);
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+
+        return dataSet;
+    }
+
+    private async Task<DataSetVersion> CreateDataSetVersion(DataSet dataSet,
+        Guid releaseFileId,
+        CancellationToken cancellationToken)
+    {
+        var dataSetVersion = new DataSetVersion
+        {
+            ReleaseFileId = releaseFileId,
+            DataSetId = dataSet.Id,
+            Status = DataSetVersionStatus.Processing,
+            Imports =
+            [
+                new DataSetVersionImport
+                {
+                    InstanceId = Guid.NewGuid(), Stage = DataSetVersionImportStage.Created
+                }
+            ],
+            Notes = "",
+            VersionMajor = dataSet.LatestLiveVersion?.VersionMajor ?? 1,
+            VersionMinor = dataSet.LatestLiveVersion?.VersionMinor + 1 ?? 0
+        };
+
+        dataSet.Versions.Add(dataSetVersion);
+        dataSet.LatestDraftVersion = dataSetVersion;
+
+        publicDataDbContext.DataSets.Update(dataSet);
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+
+        return dataSetVersion;
+    }
+
+    private async Task<Either<ActionResult, DataSet>> GetOrCreateDataSet(Guid? dataSetId,
+        Guid releaseFileId,
+        CancellationToken cancellationToken)
+    {
+        return dataSetId.HasValue
+            ? await GetDataSet(dataSetId.Value, cancellationToken)
+            : await CreateDataSet(releaseFileId, cancellationToken);
+    }
+
+    private async Task<Either<ActionResult, DataSet>> GetDataSet(Guid dataSetId,
+        CancellationToken cancellationToken)
+    {
+        return await publicDataDbContext.DataSets
+            .Include(ds => ds.LatestLiveVersion)
+            .FirstOrNotFoundAsync(ds => ds.Id == dataSetId, cancellationToken);
+    }
+
+    private async Task UpdateFilePublicDataSetVersionId(Guid releaseFileId,
+        Guid dataSetVersionId,
+        CancellationToken cancellationToken)
+    {
+        var file = await contentDbContext.ReleaseFiles
+            .Where(rf => rf.Id == releaseFileId)
+            .Select(rf => rf.File)
+            .FirstAsync(cancellationToken);
+
+        file.PublicDataSetVersionId = dataSetVersionId;
+
+        await contentDbContext.SaveChangesAsync(cancellationToken);
+    }
+
+    private static ErrorViewModel CreateError(string code,
+        string message,
+        Guid? dataSetId = null,
+        Guid? releaseFileId = null)
+    {
+        return new ErrorViewModel
+        {
+            Code = code,
+            Message = message,
```

Comment:
Ah yeah I was already trying to avoid the global errors by adding `releaseFieldId` as the path and the only reason I didn't was because this method creating the errors is separated far away from the original request type which is only in the function trigger.

I will make the change to pass the request type through to the service as suggested, thanks.